### ISSUE TITLE
request, new "exp" mode for sliders/nbx

### ISSUE THREAD

**Issue by porres** [pure-data/pure-data#1647]
I'd like to propose a second 'exponential' mode for iemguis besides the 'log' mode.

This mode would allow you to set your exponential factor (which is fixed in 'log' mode). As a reference, here's how I do it in one of my objects [rescale]. The exponential factor 'x' is like raising to the power of 'x'. So a factor of 2 is like raising to the square. The input range is remaped to the range from 0 to 1 and then this factor is applied. Negative values are also possible and it mirrors the curve vertically. the screenshot below shows this for a range from 0 to 1 but, again, the idea is to allow any range, so 0 here actually means "minimum" and 1 is "maximum".

Well, this seems quite more versatile than the current log mode and I really like it, what do you think?

![Screen Shot 2022-04-24 at 19 53 55](https://user-images.githubusercontent.com/5360142/165000292-d7fd4066-8661-4111-9c02-1c0f501e6c28.png)

**Comment by porres** [comment]
Of course this would requiere a redesign of properties window with a slot for the exponential facor. And a factor of "1" is linear and we already have. So maybe we can just add this factor number slot, which gets grayed out if it's in 'log' mode.

A new 'exp' message could define the exponential factor, which would also be invalide/ignored if in 'log' mode.

**Comment by umlaeute** [comment]
it seems counterintuitive to have to modes that do the same thing, with one being a generalized form of the other.
it might be better to extend the existing mode to accept an exponential factor.

in any case: this is going to be a problem with backwards compatibility (as in: the "save" representation of iemguis is already hell)

**Comment by porres** [comment]
> the "save" representation of iemguis is already hell

oh yeah.... :/

**Comment by Ant1r** [comment]
the current implementation uses an int field called `x_lin0_log1`, which is saved to file.
it could be changed to a float, which would drive a log/exp curve as @porres suggests, in a way that "1" value corresponds to the current "log" behavior. The "save" format would be the same as now (and backward compatibility ensured).

**Comment by umlaeute** [comment]
that would be an option.

personally, i would prefer if the log mode would be dropped completely from sliders, and instead advise the user to do some post-processing to the scale they like.

most of the iemguis suffer severely from feature-creep. but of course it's far too late now.

**Comment by porres** [comment]
> most of the iemguis suffer severely from feature-creep.

Agree. I can also point things like Moving objects (delta/pos), getting location of canvases, labels (a separate 'comment-gui' would be nice instead) and init values.

> personally, i would prefer it if the log mode would be dropped completely from sliders, and instead advise the user to do some post-processing to the scale they like.

In my library, [rescale] serves that purpose, but then, post-processing and adjusting to your own log scale isn't all that trivial and since we have this option for sliders I think it makes sense to provide a new and more versatile function and keep the older option out of focus.

Anyway, I can eventually propose some code changes for this but I guess need help to redesign the properties windows - I still struggle a lot with tcl/tk
